utils/build_swift/presets.py

```diff
@@ -0,0 +1,247 @@
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
+#
+# See https://swift.org/LICENSE.txt for license information
+# See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+
+
+"""
+Swift preset parsing and handling functionality.
+"""
+
+
+from collections import namedtuple
+from copy import copy
+
+try:
+    # Python 3
+    import configparser
+    from io import StringIO
+except ImportError:
+    import ConfigParser as configparser
+    from StringIO import StringIO
+
+
+__all__ = [
+    'Preset',
+    'PresetParser',
+
+    'InterpolationError',
+    'MissingOptionError',
+    'PresetNotFoundError',
+    'UnparsedFilesError',
+]
+
+
+# -----------------------------------------------------------------------------
+
+_RawPreset = namedtuple('_RawPreset', ['name', 'mixins', 'args'])
+
+
+def _interpolate_string(string, values):
+    if string is None:
+        return string
+
+    try:
+        return string % values
+    except KeyError as e:
+        raise InterpolationError(e.message)
+
+
+def _remove_prefix(string, prefix):
+    if string.startswith(prefix):
+        return string[len(prefix):]
+    return string
+
+
+# -----------------------------------------------------------------------------
+
+class InterpolationError(Exception):
+    """Error indicating a filaure while interpolating variables in preset
+    argument values.
+    """
+
+    pass
+
+
+class MissingOptionError(Exception):
+    """Error indicating a missing option while parsing presets.
+    """
+
+    pass
+
+
+class PresetNotFoundError(Exception):
+    """Error indicating failure when attempting to get a preset.
+    """
+
+    pass
+
+
+class UnparsedFilesError(Exception):
+    """Error indicating failure when parsing preset files.
+    """
+
+    pass
+
+
+class Preset(namedtuple('Preset', ['name', 'args'])):
+    """Container class used to wrap preset names and expanded argument lists.
+    """
+
+    # Keeps memory costs low according to the docs
+    __slots__ = ()
+
+    def format_args(self):
+        """Format argument pairs for use in the command line.
+        """
+
+        args = []
+        for (name, value) in self.args:
+            if value is None:
+                args.append(name)
+            else:
+                args.append('{}={}'.format(name, value))
+
+        return args
+
+
+# -----------------------------------------------------------------------------
+
+class PresetParser(object):
+    """Parser class used to read and manipulate Swift preset files.
+    """
+
+    _PRESET_PREFIX = 'preset: '
+
+    def __init__(self):
+        self._parser = configparser.RawConfigParser(allow_no_value=True)
+        self._presets = {}
+
+    def _parse_raw_preset(self, section):
+        preset_name = _remove_prefix(section, PresetParser._PRESET_PREFIX)
```

Comment:
`SwiftBuildSupport.py` used to ignore non-prefixed sections.
Are we going to make `preset: ` optional?

Reply:
For the time-being we are going to maintain feature parity with the old implementation. In the future we might consider changing the format some.